<a href="https://colab.research.google.com/github/spasatel13/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/DS21_W1_D2_Solutions_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Make Features

## Objectives

- Understand **the purpose of feature engineering**
- **Work with strings** in Pandas
- **Modify or create columns** of a dataframe using the `.apply()` function
- Work with **dates and times** in Pandas

## Feature Engineering

> #### Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas, which is what we'll be practicing today!

#### What are features?

#### Feature Extraction

#### Feature Construction

#### The Process

### Getting Started

Columns of a dataframe each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [73]:
import pandas as pd

In [74]:
# Pandas Display Options:
pd.set_option('display.max_rows', 150)
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows',130)
pd.set_option('display.max_columns',85)

In [75]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

In [ ]:
!curl https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv

In [77]:
print(df.shape)
df.head()
df

(1460, 81)


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1            Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2            Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3            Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4            Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
...          ...       ...       ...       ...          ...        ...   
1455         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
1456         Lvl    AllPub    Inside       Gtl       NWAmes       Norm   
1457         Lvl    AllPub    Inside       Gtl      Crawfor       Norm   
1458         Lvl    AllPub    Inside       Gtl        NAmes       Norm   
1459         Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     2Story            7            5       2003   
1          Norm     1Fam     1Story            6            8       1976   
2          Norm     1Fam     2Story            7            5       2001   
3          Norm     1Fam     2Story            7            5       1915   
4          Norm     1Fam     2Story            8            5       2000   
...         ...      ...        ...          ...          ...        ...   
1455       Norm     1Fam     2Story            6            5       1999   
1456       Norm     1Fam     1Story            6            6       1978   
1457       Norm     1Fam     2Story            7            9       1941   
1458       Norm     1Fam     1Story            5            6       1950   
1459       Norm     1Fam     1Story            5            6       1965   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1             1976     Gable  CompShg     MetalSd     MetalSd       None   
2             2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3             1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4             2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
...            ...       ...      ...         ...         ...        ...   
1455          2000     Gable  CompShg     VinylSd     VinylSd       None   
1456          1988     Gable  CompShg     Plywood     Plywood      Stone   
1457          2006     Gable  CompShg     CemntBd     CmentBd       None   
1458          1996       Hip  CompShg     MetalSd     MetalSd       None   
1459          1965     Gable  CompShg     HdBoard     HdBoard       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0          196.0        Gd        TA      PConc       Gd       TA   
1            0.0        TA        TA     CBlock       Gd       TA   
2          162.0        Gd        TA      PConc       Gd       TA   
3            0.0        TA        TA     BrkTil       TA       Gd   

### Analyzing Column Data Types

In [78]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [79]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!
df['BedroomAbvGr']

0       3
1       3
2       3
3       3
4       4
       ..
1455    3
1456    3
1457    4
1458    2
1459    3
Name: BedroomAbvGr, Length: 1460, dtype: int64

Some columns hold float values like the `LotFrontage` column.

In [80]:
# Look at the first ten rows of the `BedroomAbvGr` column.
df['LotFrontage']

0       65.0
1       80.0
2       68.0
3       60.0
4       84.0
        ... 
1455    62.0
1456    85.0
1457    66.0
1458    68.0
1459    75.0
Name: LotFrontage, Length: 1460, dtype: float64

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [ ]:
df['LotFrontage'].value_counts()

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [ ]:
# NaN is the most common value in this column. What is a NaN
df['LotFrontage'].value_counts(dropna=False)

In [71]:
df.isnull().sum()

loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
CleanedRate                   0
term                          0
                          ...  
settlement_amount        127441
settlement_percentage    127441
settlement_term          127441
cleaned_int_rate2             0
cleanedList                   0
Length: 143, dtype: int64

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [83]:
import numpy as np

# What is the datatype of NaN?
type(np.NaN)

float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

In [84]:
df['BedroomAbvGr']

0       3
1       3
2       3
3       3
4       4
       ..
1455    3
1456    3
1457    4
1458    2
1459    3
Name: BedroomAbvGr, Length: 1460, dtype: int64

In [85]:
# dot syntax does not work in every circumstance
df.BedroomAbvGr

0       3
1       3
2       3
3       3
4       4
       ..
1455    3
1456    3
1457    4
1458    2
1459    3
Name: BedroomAbvGr, Length: 1460, dtype: int64

In [86]:
df['BedroomAbvGr'][0] = np.NaN

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [87]:
df['BedroomAbvGr'] # all the ints were cast into float

0       NaN
1       3.0
2       3.0
3       3.0
4       4.0
       ... 
1455    3.0
1456    3.0
1457    4.0
1458    2.0
1459    3.0
Name: BedroomAbvGr, Length: 1460, dtype: float64

### Making New Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice`


In [88]:
# get a single column
df['TotalBsmtSF']

0        856
1       1262
2        920
3        756
4       1145
        ... 
1455     953
1456    1542
1457    1152
1458    1078
1459    1256
Name: TotalBsmtSF, Length: 1460, dtype: int64

In [ ]:
# Get more than one column, then pass in a list of column headers

df[['TotalBsmtSF', '1stFlrSF']]

In [ ]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()

small_df.head()

In [91]:
small_df.dtypes

TotalBsmtSF    int64
1stFlrSF       int64
2ndFlrSF       int64
SalePrice      int64
dtype: object

### Creating New Columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [ ]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

# Using bracket syntax to make a new 'TotalSquareFootage' column
#insert at the end
small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']

# insert at a certain position
small_df.insert(1,'13',small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF'])

#small_df.drop('TotalSquareFootage13', axis=1,inplace=True)
small_df['Price per Square Foot']=small_df['SalePrice']/small_df['TotalSquareFootage']
small_df.head()

In [93]:
small_df.drop('13',axis=1, inplace=True)
small_df.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   Price per Square Foot  
0              81.254871  
1              71.909667  
2              82.594235  
3              56.611403  
4              74.783129

In [ ]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

small_df['price_per_square_ft'] = small_df['SalePrice'] / small_df['TotalSquareFootage']

small_df.head()

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Working with Strings in Pandas


So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Let's Load a New Dataset

In [95]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-10-05 02:51:38--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 52.27.223.142, 34.211.109.164, 54.68.156.83
Connecting to resources.lendingclub.com (resources.lendingclub.com)|52.27.223.142|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip.1’

LoanStats_2018Q4.cs     [                 <=>]  22.30M  2.17MB/s    in 10s     

2020-10-05 02:51:49 (2.15 MB/s) - ‘LoanStats_2018Q4.csv.zip.1’ saved [23387713]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [96]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
replace LoanStats_2018Q4.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [97]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [98]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","2287.33","2287.33","4364.28","4364.28","3312.67","1051.61","0.0","0.0","0.0","Sep-2020","190.21","Oct-2020","Sep-2020","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [ ]:
# Read in the CSV
# set the column headers to the first row
df = pd.read_csv('LoanStats_2018Q4.csv', header=1)

print(df.shape)
df.head()

In [ ]:
df.dtypes

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [ ]:
# We can fix the header problem by using the 'skiprows' parameter

df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)

print(df.shape)
df.head()

Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [ ]:
# Sum null values by column and sort from least to greatest

df.isnull().sum().sort_values()

In [102]:
df.tail()

id  member_id  \
128375                                               NaN        NaN   
128376                                               NaN        NaN   
128377                                               NaN        NaN   
128378  Total amount funded in policy code 1: 2050322100        NaN   
128379   Total amount funded in policy code 2: 819980372        NaN   

        loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
128375     5000.0       5000.0           5000.0   36 months   13.56%   
128376    10000.0      10000.0           9750.0   36 months   11.06%   
128377    10000.0      10000.0          10000.0   36 months   16.91%   
128378        NaN          NaN              NaN         NaN      NaN   
128379        NaN          NaN              NaN         NaN      NaN   

        installment grade sub_grade             emp_title emp_length  \
128375       169.83     C        C1          Payoff Clerk  10+ years   
128376       327.68     B        B3                   NaN        NaN   
128377       356.08     C        C5  Key Accounts Manager    2 years   
128378          NaN   NaN       NaN                   NaN        NaN   
128379          NaN   NaN       NaN                   NaN        NaN   

       home_ownership  annual_inc verification_status   issue_d loan_status  \
128375       MORTGAGE     35360.0        Not Verified  Oct-2018     Current   
128376           RENT     44400.0     Source Verified  Oct-2018     Current   
128377           RENT     80000.0        Not Verified  Oct-2018     Current   
128378            NaN         NaN                 NaN       NaN         NaN   
128379            NaN         NaN                 NaN       NaN         NaN   

       pymnt_plan  url  desc             purpose                    title  \
128375          n  NaN   NaN  debt_consolidation       Debt consolidation   
128376          n  NaN   NaN         credit_card  Credit card refinancing   
128377          n  NaN   NaN               other                    Other   
128378        NaN  NaN   NaN                 NaN                      NaN   
128379        NaN  NaN   NaN                 NaN                      NaN   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128375    381xx         TN  11.30          1.0         Jun-2006   
128376    980xx         WA  11.78          0.0         Oct-2008   
128377    021xx         MA  17.72          1.0         Sep-2006   
128378      NaN        NaN    NaN          NaN              NaN   
128379      NaN        NaN    NaN          NaN              NaN   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128375             0.0                    21.0                     NaN   
128376             2.0                    40.0                     NaN   
128377             0.0                    14.0                     NaN   
128378             NaN                     NaN                     NaN   
128379             NaN                     NaN                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128375       9.0      0.0     2597.0      27.3%       15.0   
128376      15.0      0.0     6269.0      13.1%       25.0   
128377      17.0      0.0     1942.0      30.8%       31.0   
128378       NaN      NaN        NaN        NaN        NaN   
128379       NaN      NaN        NaN        NaN        NaN   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128375                   f    2042.26        2042.26      3902.32   
128376                   f    3996.89        3896.97      7521.28   
128377                   w    4202.86        4202.86      8180.45   
128378                 NaN        NaN            NaN          NaN   
128379                 NaN        NaN            NaN          NaN   

        total_pymnt_inv  total_rec_prncp  total_rec_int  ...  \
128375          3902.32          2957.74         944.58  ...   
128376          7333.25          6003.11        1518.17  ..

In [ ]:
# Address the extra NaNs in each column by skipping the footer as well.

df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1, skipfooter=2, engine='python')

print(df.shape)
df.head()

In [35]:
df.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
128373 NaN        NaN      23000        23000          23000.0   36 months   
128374 NaN        NaN      10000        10000          10000.0   36 months   
128375 NaN        NaN       5000         5000           5000.0   36 months   
128376 NaN        NaN      10000        10000           9750.0   36 months   
128377 NaN        NaN      10000        10000          10000.0   36 months   

       int_rate  installment grade sub_grade             emp_title emp_length  \
128373   15.02%       797.53     C        C3        Tax Consultant  10+ years   
128374   15.02%       346.76     C        C3        security guard    5 years   
128375   13.56%       169.83     C        C1          Payoff Clerk  10+ years   
128376   11.06%       327.68     B        B3                   NaN        NaN   
128377   16.91%       356.08     C        C5  Key Accounts Manager    2 years   

       home_ownership  annual_inc verification_status   issue_d  loan_status  \
128373       MORTGAGE     75000.0     Source Verified  Oct-2018  Charged Off   
128374       MORTGAGE     38000.0        Not Verified  Oct-2018      Current   
128375       MORTGAGE     35360.0        Not Verified  Oct-2018      Current   
128376           RENT     44400.0     Source Verified  Oct-2018      Current   
128377           RENT     80000.0        Not Verified  Oct-2018      Current   

       pymnt_plan  url  desc             purpose                    title  \
128373          n  NaN   NaN  debt_consolidation       Debt consolidation   
128374          n  NaN   NaN  debt_consolidation       Debt consolidation   
128375          n  NaN   NaN  debt_consolidation       Debt consolidation   
128376          n  NaN   NaN         credit_card  Credit card refinancing   
128377          n  NaN   NaN               other                    Other   

       zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
128373    352xx         AL  20.95            1         Aug-1985   
128374    443xx         OH  13.16            3         Jul-1982   
128375    381xx         TN  11.30            1         Jun-2006   
128376    980xx         WA  11.78            0         Oct-2008   
128377    021xx         MA  17.72            1         Sep-2006   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
128373               2                    22.0                     NaN   
128374               0                     6.0                     NaN   
128375               0                    21.0                     NaN   
128376               2                    40.0                     NaN   
128377               0                    14.0                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
128373        12        0      22465      43.6%         28   
128374        11        0       5634      37.1%         16   
128375         9        0       2597      27.3%         15   
128376        15        0       6269      13.1%         25   
128377        17        0       1942      30.8%         31   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
128373                   w       0.00           0.00      1547.08   
128374                   w    4136.11        4136.11      7967.14   
128375                   f    2042.26        2042.26      3902.32   
128376                   f    3996.89        3896.97      7521.28   
128377                   w    4202.86        4202.86      8180.45   

        total_pymnt_inv  total_rec_prncp  total_rec_int  ...  \
128373          1547.08          1025.67         521.41  ...   
128374          7967.14          5863.89        2103.25  ...   
128375          3902.32          2957.74         944.58  ...   
128376          7333.25          6003.11        1518.17  ...   
128377          8180.45          5797.14        2383.31  ...   

        num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  \
128373                 

In [104]:
df.isnull().sum().sort_values()

inq_fi                           0
mo_sin_old_rev_tl_op             0
delinq_amnt                      0
chargeoff_within_12_mths         0
acc_open_past_24mths             0
                             ...  
settlement_percentage       127441
desc                        128378
url                         128378
member_id                   128378
id                          128378
Length: 144, dtype: int64

In [105]:
df.dtypes

id                       float64
member_id                float64
loan_amnt                  int64
funded_amnt                int64
funded_amnt_inv          float64
                          ...   
settlement_status         object
settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
Length: 144, dtype: object

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [ ]:
df.head()

In [107]:
df = df.drop(['id', 'member_id', 'desc', 'url'], axis=1)

In [40]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4      Educator     2 years           RENT     46500.0   
4     B        B5   Tire builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  ...  num_tl_op_past_12m  \
0         0.0                      0.0     Sep-2020  ...                   1   
1         0.0                      0.0     Sep-2020  ...                   4   
2         0.0                      0.0     Feb-2019  ...                   1   
3         0.0                      0.0     Jul-2020  ...                   9   
4         0.0                      0.0     Aug-20

### Cleaning Up the `int_rate` Column

When we're preparing a dataset for a machine learning model we typically don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [108]:
# Look at the first 10 values of the int_rate column

df['int_rate'].head(10)

0     11.31%
1     18.94%
2      7.56%
3     11.80%
4     12.98%
5     10.33%
6     19.92%
7     23.40%
8     15.02%
9     12.98%
Name: int_rate, dtype: object

In [109]:
# Look at a specific value from the int_rate column

df['int_rate'][0]

' 11.31%'

In [110]:
df['int_rate'][2]

'  7.56%'

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [111]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them

int_rate = ' 14.47%'

my_num = 6

In [112]:
# my_num.strip()

In [113]:
int_rate.strip()

'14.47%'

In [114]:
int_rate = ' 14.47%'

int_rate.strip().strip('%')

'14.47'

In [115]:
# "Chaining"
# df.isnull().sum()

int_rate.strip().strip('%')

'14.47'

In [116]:
# "Cast" the string value to a float
# "cast" -> Change something's data type
# This is the line of code that we're after! ->
float(int_rate.strip().strip('%'))

14.47

In [117]:
type(14.47)

float

In [118]:
type(' 14.47%')

str

In [119]:
type(float(int_rate.strip().strip('%')))

float

### Write a function to make our solution reusable!

In [120]:
df['int_rate'].head()

0     11.31%
1     18.94%
2      7.56%
3     11.80%
4     12.98%
Name: int_rate, dtype: object

In [121]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.
def int_rate_to_float(cell_contents):
  return float(cell_contents.strip().strip('%'))

In [122]:
# Test out our function by calling it on our example
int_rate_to_float(int_rate)

14.47

In [123]:
# is the data type correct?
type(int_rate_to_float(int_rate))

float

### `.apply()` our solution to every cell in a column

In [124]:
df['int_rate'].apply(int_rate_to_float)

0         11.31
1         18.94
2          7.56
3         11.80
4         12.98
          ...  
128373    15.02
128374    15.02
128375    13.56
128376    11.06
128377    16.91
Name: int_rate, Length: 128378, dtype: float64

In [125]:
# pass in *only* the name of the function, don't call it. 
# This works because we know the function works on every item in the column
# so I can siply "apply" it to the entire column
df['int_rate_cleaned'] = df['int_rate'].apply(int_rate_to_float)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4      Educator     2 years           RENT     46500.0   
4     B        B5   Tire builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  ...  pct_tl_nvr_dlq  \
0         0.0                      0.0     Sep-2020  ...           100.0   
1         0.0                      0.0     Sep-2020  ...            93.3   
2         0.0                      0.0     Feb-2019  ...            94.7   
3         0.0                      0.0     Jul-2020  ...            93.3   
4         0.0                      0.0     Aug-2020  ...            8

In [126]:
# What type of data is held in our new column?

# Look at the datatypes of the last 5 columns

df.dtypes.tail()

settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
int_rate_cleaned         float64
dtype: object

### Alternative Approach (not using `.apply()`)

In [127]:
clean_int_rates = []

for int_rate in df['int_rate']:
  clean_int_rates.append(int_rate_to_float(int_rate))

clean_int_rates

# Take that list and turn it into a datafram column "Series"

df['cleaned_int_rate2'] = pd.Series(clean_int_rates)
df['cleaded_int_rate3'] = pd.Series(clean_int_rates)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4      Educator     2 years           RENT     46500.0   
4     B        B5   Tire builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  ...  \
0         0.0                      0.0     Sep-2020  ...   
1         0.0                      0.0     Sep-2020  ...   
2         0.0                      0.0     Feb-2019  ...   
3         0.0                      0.0     Jul-2020  ...   
4         0.0                      0.0     Aug-2020  ...   

   pub_rec_bankruptcies tax_liens tot_hi_cred_lim  total_bal_ex_mort  \
0               

In [128]:
#using map function to add a columnt
df['cleaned_int_rate4'] = list(map(int_rate_to_float, df['int_rate']))
df['cleaded_int_rate5'] = list(map(int_rate_to_float, df['int_rate']))

#df.drop('cleaned_int_rate2', axis=1, inplace=True)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4      Educator     2 years           RENT     46500.0   
4     B        B5   Tire builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  ...  tot_hi_cred_lim  \
0         0.0                      0.0     Sep-2020  ...           125853   
1         0.0                      0.0     Sep-2020  ...           253346   
2         0.0                      0.0     Feb-2019  ...           622183   
3         0.0                      0.0     Jul-2020  ...            50845   
4         0.0                      0.0     Aug-2020  ...        

In [129]:
df=df.drop(['int_rate_cleaned','cleaded_int_rate3','cleaned_int_rate4','cleaded_int_rate5'], axis=1)
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           5000.0   36 months   11.31%       164.43   
1      10000        10000          10000.0   36 months   18.94%       366.26   
2      20000        20000          20000.0   36 months    7.56%       622.68   
3       6500         6500           6500.0   36 months   11.80%       215.28   
4      25000        25000          25000.0   60 months   12.98%       568.58   

  grade sub_grade      emp_title emp_length home_ownership  annual_inc  \
0     B        B3       Shipping    2 years           RENT     40000.0   
1     D        D2  Especialist 1    7 years       MORTGAGE     37115.0   
2     A        A3        Teacher  10+ years       MORTGAGE    100000.0   
3     B        B4      Educator     2 years           RENT     46500.0   
4     B        B5   Tire builder  10+ years       MORTGAGE     85000.0   

  verification_status   issue_d        loan_status pymnt_plan  \
0     Source Verified  Dec-2018            Current          n   
1        Not Verified  Dec-2018            Current          n   
2        Not Verified  Dec-2018         Fully Paid          n   
3     Source Verified  Dec-2018  Late (16-30 days)          n   
4        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  ...  pct_tl_nvr_dlq  \
0         0.0                      0.0     Sep-2020  ...           100.0   
1         0.0                      0.0     Sep-2020  ...            93.3   
2         0.0                      0.0     Feb-2019  ...            94.7   
3         0.0                      0.0     Jul-2020  ...            93.3   
4         0.0                      0.0     Aug-2020  ...            8

In [63]:
df.isna().sum()

loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
term                          0
int_rate                      0
                          ...  
settlement_date          127441
settlement_amount        127441
settlement_percentage    127441
settlement_term          127441
cleaned_int_rate2             0
Length: 141, dtype: int64

In [130]:
df.shape

(128378, 141)

In [131]:
# converting int to float using `for` loop
newList=[]
for i in df['int_rate']:
  newList.append(int_rate_to_float(i))
df['cleanedList']=pd.Series(newList)
df.insert(3,'CleanedRate','newList')
df.head()

loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4      Educator     2 years           RENT   
4       568.58     B        B5   Tire builder  10+ years       MORTGAGE   

   annual_inc verification_status   issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified  Dec-2018            Current          n   
1     37115.0        Not Verified  Dec-2018            Current          n   
2    100000.0        Not Verified  Dec-2018         Fully Paid          n   
3     46500.0     Source Verified  Dec-2018  Late (16-30 days)          n   
4     85000.0        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... percent_bc_gt_75  \
0         0.0                      0.0  ...              0.0   
1         0.0                      0.0  ...             33.3   
2         0.0                      0.0  ...             20.0   
3         0.0                      0.0  ...              0.0   
4         0.0                      0.0  ...              

### Key Point(s)

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

## Modifying/Creating Columns with Pandas

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the `emp_title` "Employment Title" column!

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [149]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False).head(120)

NaN                          20943
Teacher                       2089
Manager                       1773
Registered Nurse               950
Driver                         924
RN                             726
Supervisor                     697
Sales                          580
Project Manager                526
General Manager                523
Office Manager                 521
Owner                          420
Director                       402
Truck Driver                   387
Operations Manager             387
Nurse                          326
Engineer                       325
Sales Manager                  304
manager                        301
Supervisor                     270
Administrative Assistant       269
Accountant                     268
Server                         265
Vice President                 261
Mechanic                       258
Account Manager                254
Police Officer                 252
teacher                        249
Technician          

In [137]:
# How many different unique employment titles are there currently?
df['emp_title'].nunique()

43881

In [162]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20943

What are some possible reasons as to why a person's employment title may have not been provided?

In [150]:
# Create some examples that represent the cases that we want to clean up
examples = ['owner', 'Supervisor', ' Project Manager', 'TEACHER', np.NaN]

In [151]:
# Write a function to clean up these use cases and increase uniformity.
def clean_emp_title(title):
  # type checking
  if isinstance(title, str):
    # Give "title" casing and remove leading/trailing whitespace
    return title.strip().title()
  else:
    return "Unknown"

In [152]:
for example in examples:
  # print things that you want to force to be shown in the notebook output
  print(clean_emp_title(example))

Owner
Supervisor
Project Manager
Teacher
Unknown


In [153]:
# The value of the last thing to be returned gets printed out
clean_emp_title("owner")
clean_emp_title("TEACHER")

'Teacher'

In [154]:
# if the last thing to be returned gets saved to a variable
# then it won't get printed out
new_title = clean_emp_title("TEACHER")

In [155]:
# if I include just a variable name, it will get printed out

new_title

'Teacher'

In [167]:
df['emp_title_cleaned'] = df['emp_title'].apply(clean_emp_title)
df['new_emp2']=(list(map(clean_emp_title,df['emp_title'])))

In [164]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4      Educator     2 years           RENT   
4       568.58     B        B5   Tire builder  10+ years       MORTGAGE   

   annual_inc verification_status   issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified  Dec-2018            Current          n   
1     37115.0        Not Verified  Dec-2018            Current          n   
2    100000.0        Not Verified  Dec-2018         Fully Paid          n   
3     46500.0     Source Verified  Dec-2018  Late (16-30 days)          n   
4     85000.0        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... tot_hi_cred_lim  \
0         0.0                      0.0  ...          125853   
1         0.0                      0.0  ...          253346   
2         0.0                      0.0  ...          622183   
3         0.0                      0.0  ...           50845   
4         0.0                      0.0  ...          275691   

In [168]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.

# Combining a for loop and adding items to a list into a single line

[clean_emp_title(title) for title in examples]

['Owner', 'Supervisor', 'Project Manager', 'Teacher', 'Unknown']

In [169]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column

df['emp_title'] = df['emp_title'].apply(clean_emp_title)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4       Educator    2 years           RENT   
4       568.58     B        B5   Tire Builder  10+ years       MORTGAGE   

   annual_inc verification_status   issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified  Dec-2018            Current          n   
1     37115.0        Not Verified  Dec-2018            Current          n   
2    100000.0        Not Verified  Dec-2018         Fully Paid          n   
3     46500.0     Source Verified  Dec-2018  Late (16-30 days)          n   
4     85000.0        Not Verified  Dec-2018         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... tot_hi_cred_lim  \
0         0.0                      0.0  ...          125853   
1         0.0                      0.0  ...          253346   
2         0.0                      0.0  ...          622183   
3         0.0                      0.0  ...           50845   
4         0.0                      0.0  ...          275691   

We can use the same code as we did earlier to see how much progress was made.


In [170]:
# Look at the top 20 employment titles

df['emp_title_cleaned'].value_counts(dropna=False).sort_values(ascending=False).head(20)

Unknown               20943
Teacher                2556
Manager                2395
Registered Nurse       1416
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              419
Engineer                412
Name: emp_title_cleaned, dtype: int64

In [171]:
# How many different unique employment titles are there currently?
df['emp_title_cleaned'].nunique()

34895

In [172]:
# How often is the employment_title null (NaN)?'

df['emp_title_cleaned'].isnull().sum()

0

### Key Point(s)

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

## Working with Dates & Times in Pandas

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

Pandas documentation:
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [174]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Lets look at the string format of the `issue_d` column

In [178]:
df['issue_d'].head()

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
Name: issue_d, dtype: object

In [179]:
df['issue_d'][0]

'Dec-2018'

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [180]:
pd.to_datetime(df['issue_d'])

0        2018-12-01
1        2018-12-01
2        2018-12-01
3        2018-12-01
4        2018-12-01
            ...    
128373   2018-10-01
128374   2018-10-01
128375   2018-10-01
128376   2018-10-01
128377   2018-10-01
Name: issue_d, Length: 128378, dtype: datetime64[ns]

In [187]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4       Educator    2 years           RENT   
4       568.58     B        B5   Tire Builder  10+ years       MORTGAGE   

   annual_inc verification_status    issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified 2018-12-01            Current          n   
1     37115.0        Not Verified 2018-12-01            Current          n   
2    100000.0        Not Verified 2018-12-01         Fully Paid          n   
3     46500.0     Source Verified 2018-12-01  Late (16-30 days)          n   
4     85000.0        Not Verified 2018-12-01         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... tot_hi_cred_lim  \
0         0.0                      0.0  ...          125853   
1         0.0                      0.0  ...          253346   
2         0.0                      0.0  ...          622183   
3         0.0                      0.0  ...           50845   
4         0.0                      0.0  ...          275

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [183]:
df['issue_d'].head()

0   2018-12-01
1   2018-12-01
2   2018-12-01
3   2018-12-01
4   2018-12-01
Name: issue_d, dtype: datetime64[ns]

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [188]:
df['issue_d'][0]

Timestamp('2018-12-01 00:00:00')

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [189]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128373    2018
128374    2018
128375    2018
128376    2018
128377    2018
Name: issue_d, Length: 128378, dtype: int64

Now the month.

In [190]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128373    10
128374    10
128375    10
128376    10
128377    10
Name: issue_d, Length: 128378, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [191]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

In [192]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4       Educator    2 years           RENT   
4       568.58     B        B5   Tire Builder  10+ years       MORTGAGE   

   annual_inc verification_status    issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified 2018-12-01            Current          n   
1     37115.0        Not Verified 2018-12-01            Current          n   
2    100000.0        Not Verified 2018-12-01         Fully Paid          n   
3     46500.0     Source Verified 2018-12-01  Late (16-30 days)          n   
4     85000.0        Not Verified 2018-12-01         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Oct-2007               0                     NaN   
1            0         Jan-2002               0                    38.0   
2            0         Feb-1999               0                    48.0   
3            0         Aug-2011               1                    61.0   
4            1         Oct-1993               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... total_bc_limit  \
0         0.0                      0.0  ...           3600   
1         0.0                      0.0  ...           5600   
2         0.0                      0.0  ...          85100   
3         0.0                      0.0  ...          13300   
4         0.0                      0.0  ...          43900   

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [193]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [194]:
(df['issue_d'] - df['earliest_cr_line'])

0         4079 days
1         6178 days
2         7243 days
3         2679 days
4         9192 days
            ...    
128373   12114 days
128374   13241 days
128375    4505 days
128376    3652 days
128377    4413 days
Length: 128378, dtype: timedelta64[ns]

In [195]:
df['credit_length_days'] = (df['issue_d'] - df['earliest_cr_line']).dt.days
df['credit_length_years'] = df['credit_length_days'] / 365
df.head()

loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4       Educator    2 years           RENT   
4       568.58     B        B5   Tire Builder  10+ years       MORTGAGE   

   annual_inc verification_status    issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified 2018-12-01            Current          n   
1     37115.0        Not Verified 2018-12-01            Current          n   
2    100000.0        Not Verified 2018-12-01         Fully Paid          n   
3     46500.0     Source Verified 2018-12-01  Late (16-30 days)          n   
4     85000.0        Not Verified 2018-12-01         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0       2007-10-01               0                     NaN   
1            0       2002-01-01               0                    38.0   
2            0       1999-02-01               0                    48.0   
3            0       2011-08-01               1                    61.0   
4            1       1993-10-01               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... revol_bal_joint  \
0         0.0                      0.0  ...             NaN   
1         0.0                      0.0  ...         19051.0   
2         0.0                      0.0  ...         43287.0   
3         0.0                      0.0  ...             NaN   
4         0.0                      0.0  ...         4871

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [196]:
df['credit_length_years'].max()

68.96164383561644

What's oldest credit history that was involved in Q4 2018? 

In [197]:
df['credit_length_days'].max()

25171

25,171 days is ~ 68.96 years of credit history!

## Key Point(s)

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.

## Some Notebook Tips 

In [198]:
# notebook cells will only print out the last thing to be returned in the cell.
df.head()
df.shape

(128378, 150)

In [199]:
# If I want both to be printed out, I need to "print" everyting above the last function call

print(df.shape)
df.head()

(128378, 150)


loan_amnt  funded_amnt  funded_amnt_inv CleanedRate        term int_rate  \
0       5000         5000           5000.0     newList   36 months   11.31%   
1      10000        10000          10000.0     newList   36 months   18.94%   
2      20000        20000          20000.0     newList   36 months    7.56%   
3       6500         6500           6500.0     newList   36 months   11.80%   
4      25000        25000          25000.0     newList   60 months   12.98%   

   installment grade sub_grade      emp_title emp_length home_ownership  \
0       164.43     B        B3       Shipping    2 years           RENT   
1       366.26     D        D2  Especialist 1    7 years       MORTGAGE   
2       622.68     A        A3        Teacher  10+ years       MORTGAGE   
3       215.28     B        B4       Educator    2 years           RENT   
4       568.58     B        B5   Tire Builder  10+ years       MORTGAGE   

   annual_inc verification_status    issue_d        loan_status pymnt_plan  \
0     40000.0     Source Verified 2018-12-01            Current          n   
1     37115.0        Not Verified 2018-12-01            Current          n   
2    100000.0        Not Verified 2018-12-01         Fully Paid          n   
3     46500.0     Source Verified 2018-12-01  Late (16-30 days)          n   
4     85000.0        Not Verified 2018-12-01         Fully Paid          n   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    284xx         NC   9.66   
1               other                    Other    891xx         NV  33.57   
2         credit_card  Credit card refinancing    982xx         WA  18.92   
3  debt_consolidation       Debt consolidation    352xx         AL  21.01   
4  debt_consolidation       Debt consolidation    356xx         AL  36.67   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0       2007-10-01               0                     NaN   
1            0       2002-01-01               0                    38.0   
2            0       1999-02-01               0                    48.0   
3            0       2011-08-01               1                    61.0   
4            1       1993-10-01               1                    10.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         8        0       1070      17.5%         27   
1                     NaN         9        0       5837      53.1%         15   
2                     NaN         9        0      25416      29.9%         19   
3                     NaN        24        0       6741      41.6%         30   
4                     NaN        14        0      26630      53.3%         27   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w    2770.45        2770.45   2956.610000   
1                   w    5147.99        5147.99   7309.420000   
2                   w       0.00           0.00  20215.792430   
3                   w    3354.53        3354.53   4083.930000   
4                   w       0.00           0.00  29563.763968   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2956.61          2229.55         727.06                 0.0   
1          7309.42          4852.01        2457.41                 0.0   
2         20215.79         20000.00         215.79                 0.0   
3          4083.93          3145.47         938.46                 0.0   
4         29563.76         25000.00        4563.76                 0.0   

   recoveries  collection_recovery_fee  ... revol_bal_joint  \
0         0.0                      0.0  ...             NaN   
1         0.0                      0.0  ...         19051.0   
2         0.0                      0.0  ...         43287.0   
3         0.0                      0.0  ...             NaN   
4         0.0                      0.0  ...         4871